### PR TITLE
Proposal for a scheduler support library

### PR DESCRIPTION
**Pull request by jopperm**
**Work in progress**, to be discussed in the design meeting on June 9.

This PR would lay the groundwork for a library of static schedulers in `circt/Support/Scheduling` (please suggest better locations).

The scheduling problems faced by HLS-like synthesis flows are often similar. The idea here is to have a hierarchy of base classes that model these variants and take care of storing and validating the problem and its solutions. Currently, only the basic problem is modeled (class `Scheduler`). Variants (e.g. "pipelined", "resource-sharing", etc.) would subclass it, add additional properties for the operations, dependences, operator types or the problem itself, and provide appropriate verification methods.

A demo client is the `-schedule-pipeline` pass, which schedules the operations inside a `staticlogic.pipeline` with unit latencies and annotates the computed start times as attributes. (This could be removed from this PR if desired -- I included this pass mainly to test the interface).

`ASAPScheduler` is a complete implementation that solves the basic problem.

Note that this is the follow-up to #1040, focussing more on the library-of-schedulers aspect, with a CIRCT-agnostic interface. The discussion of the earlier PR showed that there is no obvious way that conversion passes from a untimed to a timed dialect would want to provide the information needed for scheduling (OpInterface, Attributes, ...), so better abstractions will need to evolve over time, inside concrete synthesis flows.

### CODE REVIEW

File: include/circt/Support/Scheduling/Scheduler.h [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+/// a concrete scheduling algorithm, e.g. that there are start times available
+/// for each registered operation, and the precedence constraints as modeled by
+/// the dependences are satisfied.
+class Scheduler {
+  //===--------------------------------------------------------------------===//
+  // Aliases for API types
+  //===--------------------------------------------------------------------===//
+public:
+  /// Opaque handle for operations
+  using OperationHandle = void *;
```
Comment (mikeurbach) [inline include/circt/Support/Scheduling/Scheduler.h:75]: Could this just be an `Operation *`? That's a common way to opaquely refer to operations, and I think it works in all the APIs here.

**Review by mikeurbach — COMMENTED** [review]
Just did a quick skim over Scheduler.h. Looking forward to the ODM discussion.

File: lib/Support/Scheduling/ASAPScheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements an acyclic, as-soon-as-possible list scheduler.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Support/Scheduling/ASAPScheduler.h"
+
+#include <list>
```
Comment (lattner) [inline lib/Support/Scheduling/ASAPScheduler.cpp:14]: Super nit pick, but I'd recommend using std::deque for this.

Comment (jopperm) [reply lib/Support/Scheduling/ASAPScheduler.cpp]: Fixed, thanks!

File: include/circt/Support/Scheduling/Scheduler.h (hunk 1 above)
Comment (jopperm) [reply include/circt/Support/Scheduling/Scheduler.h:75]: Yes, I'm going to change that, I realised that just having a `void*` is terrible for emitting errors.

**Comment by jopperm** [comment]
@teqdruid I'm a bit unsure how to break down this PR into smaller pieces for review. I could do a PR for the interface first, then for the simple algorithm, and finally for the test pass. Would that be acceptable, considering that no tests could be run until the third PR?

**Comment by teqdruid** [comment]
Not being familiar with this PR, that sounds like a reasonable break down. I once had a BIG piece of pre-existing code (https://github.com/llvm/circt/issues/213) wherein many of the PRs weren't testable on their own, so there is precedent there.

Also based on yesterday's discussion, I would move it out of `circt/support/scheduling` and instead put it in `circ/scheduling`. I suspect that this is one of those areas which will grow and become a major feature.

**Comment by jopperm** [comment]
Ok, thanks! Closing this then in favor of #1268.

**Pull request closed without merging**